### PR TITLE
Fix support for cstdlib atomics under --llvm

### PR DESCRIPTION
**Pull request by ronawho**
Previously `CHPL_ATOMICS=cstdlib` with --llvm would fail at link time
with error messages like `undefined reference to 'atomic_thread_fence'`.
Here we just add `-latomic` to satisfy these references. 

Currently `-latomic` is being added manually by the compiler while assembling
llvm link args. An alternative solution would be to add it to
.lib/*/list-libraries (which the compiler already harvests link flags from)

### CODE REVIEW

**Comment by mppf** [comment]
@ronawho - I don't have a problem with updating it here but perhaps it should instead involve updating /lib/..../list-libraries which could be handled by adding it to CHPL_MAKE_BASE_LFLAGS or LIBS in the Makefiles.

Additionally I'm a bit surprised that -latomic is required at all. Are we not requesting the right C standard or something when we compile? Would it help to update Makefile.clang (possibly for just clang-included) to set -std=c11 ? There's some pretty complex logic in Makefile.clang to request the right C/C++ standards but for the "included" clang I think we control the version to some extent and probably should just request the C standard we want.

I'm talking about this:

https://github.com/chapel-lang/chapel/blob/84386079c865b8794f23a99d2a62e522de00c925/make/compiler/Makefile.clang#L58-L68

but it's unclear to me if adjusting that would help.

**Comment by dmk42** [comment]
I just experimented with `-std=` options and it did not change things.

I'm inclined to go ahead with this solution, but it's puzzling why it's needed, because using the same clang and clang++ that chpl uses, I get the same libraries.  Outside of chpl (where libatomic is not needed), the resulting ld command looks like this.
```
"/usr/bin/ld" -z relro --hash-style=gnu --eh-frame-hdr -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o a.out /usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../../x86_64-linux-gnu/crt1.o /usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../../x86_64-linux-gnu/crti.o /usr/bin/../lib/gcc/x86_64-linux-gnu/8/crtbegin.o -L/usr/bin/../lib/gcc/x86_64-linux-gnu/8 -L/usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../../x86_64-linux-gnu -L/lib/x86_64-linux-gnu -L/lib/../lib64 -L/usr/lib/x86_64-linux-gnu -L/usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../.. -L/usr/lib/llvm-7/bin/../lib -L/lib -L/usr/lib atomic.o -lstdc++ -lm -lgcc_s -lgcc -lc -lgcc_s -lgcc /usr/bin/../lib/gcc/x86_64-linux-gnu/8/crtend.o /usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../../x86_64-linux-gnu/crtn.o
```
Called from chpl, it looks like the following. Note the same libraries and .o files are involved (plus some extra ones for Chapel).
```
"/usr/bin/ld" -z relro --hash-style=gnu --eh-frame-hdr -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o tmp/hello.tmp /usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../../x86_64-linux-gnu/crt1.o /usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../../x86_64-linux-gnu/crti.o /usr/bin/../lib/gcc/x86_64-linux-gnu/8/crtbegin.o -L/home/dmk/chapel/third-party/qthread/install/linux64-x86_64-native-clang-included-none-flat-jemalloc-hwloc/lib -L/home/dmk/chapel/third-party/jemalloc/install/linux64-x86_64-native-clang-included-none/lib -L/home/dmk/chapel/third-party/hwloc/install/linux64-x86_64-native-clang-included-none-flat/lib -L/home/dmk/chapel/third-party/re2/install/linux64-x86_64-native-clang-included-none/lib -L/home/dmk/chapel/lib/linux64/clang-included/x86_64/arch-native/loc-flat/comm-none/tasks-qthreads/tmr-generic/unwind-none/mem-jemalloc/atomics-cstdlib/hwloc/re2/fs-none/lib_pic-none -L/home/dmk/chapel/third-party/hwloc/install/linux64-x86_64-native-clang-included-none-flat/lib -L/usr/bin/../lib/gcc/x86_64-linux-gnu/8 -L/usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../../x86_64-linux-gnu -L/lib/x86_64-linux-gnu -L/lib/../lib64 -L/usr/lib/x86_64-linux-gnu -L/usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../.. -L/usr/lib/llvm-7/bin/../lib -L/lib -L/usr/lib tmp/chpl__module.o /home/dmk/chapel/lib/linux64/clang-included/x86_64/arch-native/loc-flat/comm-none/tasks-qthreads/tmr-generic/unwind-none/mem-jemalloc/atomics-cstdlib/hwloc/re2/fs-none/lib_pic-none/main.o tmp/chpl_compilation_config.c.o -rpath /home/dmk/chapel/third-party/qthread/install/linux64-x86_64-native-clang-included-none-flat-jemalloc-hwloc/lib -rpath /home/dmk/chapel/third-party/jemalloc/install/linux64-x86_64-native-clang-included-none/lib -rpath /home/dmk/chapel/third-party/hwloc/install/linux64-x86_64-native-clang-included-none-flat/lib -rpath /home/dmk/chapel/third-party/re2/install/linux64-x86_64-native-clang-included-none/lib -lchpl -lgmp -lm -lgmp -ljemalloc -lchpl -lqthread -lhwloc -lm -lre2 -lpthread -lstdc++ -lm -lgcc_s -lgcc -lc -lgcc_s -lgcc /usr/bin/../lib/gcc/x86_64-linux-gnu/8/crtend.o /usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../../x86_64-linux-gnu/crtn.o
```

File: compiler/llvm/clangUtil.cpp [hunk 1]
```diff
@@ -2955,6 +2955,10 @@ void makeBinaryLLVM(void) {
     command += libName;
   }
 
+  if (strcmp(CHPL_ATOMICS, "cstdlib") == 0) {
+    command += " -latomic";
+  }
```
Comment (dmk42) [inline compiler/llvm/clangUtil.cpp]: The only problem I see is that there doesn't seem to be a libatomic on Darwin (and we also don't see a failure there), so this needs to be conditional on the OS.

**Comment by ronawho** [comment]
I was also not able to reproduce this with a standalone C program with clang/clang++ that we build.

I also ran into issues with this on darwin where libatomic is not availble (or needed), which gives me a little bit of a concern for portability. I feel a little out of my depth here, so I'd be happy to let one of you look into why libatomic seems to be needed and if there's a cleaner workaround.

File: compiler/llvm/clangUtil.cpp (hunk 1 above)
Comment (ronawho) [reply compiler/llvm/clangUtil.cpp]: Yeah, coincidentally I ran into that just as you made this comment. Fixed

**Comment by dmk42** [comment]
BTW, `list-libraries` would be equally acceptable to me, keeping in mind that the `-latomic` should not be put in there on Darwin.

**Comment by ronawho** [comment]
Possibly related -- https://clang.llvm.org/docs/Toolchain.html#libatomic-gnu: 

```
  Clang does not currently automatically link against libatomic when using
  libgcc_s. You may need to manually add ``-latomic`` to support this
  configuration when using non-native atomic operations (if you see link errors
  referring to ``__atomic_*`` functions).
```

**Comment by dmk42** [comment]
Based on this documentation reference, I'd recommend a comment indicating that we should remove the `-latomic` if we start using `compiler-rt`.

**Comment by ronawho** [comment]
> Based on this documentation reference, I'd recommend a comment indicating that we should remove the -latomic if we start using compiler-rt. (Just above the quoted paragraph, it states that the atomic routines are available in compiler-rt.)

Ah, good catch. I added a comment and I inlined the comments from that link in the commit message.

**Pull request merged**